### PR TITLE
Adds browse-basic page, Expandable/ExpandableGroup protos

### PR DESCRIPTION
**Pull request by anselmbradford**
Adds browse-basic page, Expandable/ExpandableGroup prototypes. Re-submission of https://github.com/cfpb/cfgov-refresh/pull/1134

## Additions

- Adds browse-basic prototype page and URL.
- Adds Expandable molecule.
- Adds ExpandableGroup organism.
- Adds `classList` JS polyfill.
- Adds `u-hide` utility class to hide an element.
- Adds `EventObserver.js` for adding event broadcaster capability to JS classes.

## Testing

- Visit `http://localhost:8000/browse-basic/`

## Review

- @KimberlyMunoz 
- @sebworks 
- @jimmynotjim 

## Todos

- There a a few TODOs in Expandables.js source code.

### CODE REVIEW

File: cfgov/unprocessed/css/cf-enhancements.less [hunk 1]
```diff
@@ -520,6 +520,29 @@ tbody {
...
+      codenotes:
+        - .u-hide;
+      notes:
+        - "Use this class to hide something,
+          perhaps to be shown using JavaScript."
+  tags:
+    - cf-core
+*/
+
+.u-hide {
```
Comment (KimberlyMunoz) [inline cfgov/unprocessed/css/cf-enhancements.less]: Isn't this a dupe of `.u-hidden`?

File: cfgov/unprocessed/js/molecules/Expandable.js [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+  // when used in an ExpandableGroup, otherwise it can be the parent container.
+  var _dom = element.classList.contains( BASE_CLASS ) ?
+             element : element.querySelector( '.' + BASE_CLASS );
+  var _target = _dom.querySelector( '.' + BASE_CLASS + '_target' );
+  var _content = _dom.querySelector( '.' + BASE_CLASS + '_content' );
+  var _contentAnimated =
+    _content.querySelector( '.' + BASE_CLASS + '_content-animated' );
+  var _link = _dom.querySelector( '.' + BASE_CLASS + '_link' );
+
+  var _state = 0;
```
Comment (sebworks) [inline cfgov/unprocessed/js/molecules/Expandable.js]: should you be using the constant here?

File: cfgov/jinja2/v1/_includes/molecules/expandable.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+   ========================================================================== #}
+
+{% macro render(label="", settings) %}
+{# Add the __expanded modifier class by default so all expandables are open
+   when JS is switched off, but add a data-js-hook to tell JS to remove it. #}
+<div class="m-expandable
+            m-expandable__expanded
+            {{ 'm-expandable__borders' if settings.is_bordered else '' }}
+            {{ 'm-expandable__midtone' if settings.is_midtone else '' }}"
+     {{ 'data-js-hook=expanded' if settings.is_expanded else '' }}>
```
Comment (sebworks) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: Instead of calling this `data-js-hook` (which is really used to query the dom), does it make sense to call it `data-state=expanded`?

File: cfgov/unprocessed/css/cf-enhancements.less (hunk 1 above)
Comment (anselmbradford) [inline cfgov/unprocessed/css/cf-enhancements.less]: Is there a `.u-hidden`? I'm only aware of `u-visually-hidden`.

Comment (KimberlyMunoz) [inline cfgov/unprocessed/css/cf-enhancements.less]: Yep. Down in cf-enhancements, there's that an a class to only hide when JS is enabled.

File: cfgov/unprocessed/js/molecules/Expandable.js (hunk 2 above)
Comment (anselmbradford) [inline cfgov/unprocessed/js/molecules/Expandable.js]: Like `var _state = COLLAPSED`. Yeah that makes sense.

File: cfgov/unprocessed/js/molecules/Expandable.js [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+   * Set the height to zero.
+   */
+  function _setMinHeight() {
+    _content.style.height = '0';
+  }
+
+  /**
+   * @returns {boolean} Whether Expandable is in a collapsed state.
+   */
+  function _isCollapsed() {
```
Comment (sebworks) [inline cfgov/unprocessed/js/molecules/Expandable.js]: Why not just use Object.defineProperty set the state?

File: cfgov/unprocessed/css/cf-enhancements.less (hunk 1 above)
Comment (anselmbradford) [inline cfgov/unprocessed/css/cf-enhancements.less]: Ahh, haha, I'd scoured CF but didn't look in cf-enhancements. Yeah this is a dupe then.

Comment (anselmbradford) [inline cfgov/unprocessed/css/cf-enhancements.less]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/edaeaa74656fba1628f0b9706b2178d9f4a49fed

Also removed some redundant changelog entries I found.

File: cfgov/unprocessed/js/molecules/Expandable.js (hunk 2 above)
Comment (anselmbradford) [inline cfgov/unprocessed/js/molecules/Expandable.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/2d403e1899097e831033443b174af07d2a389545

File: cfgov/jinja2/v1/_includes/molecules/expandable.html (hunk 3 above)
Comment (anselmbradford) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: I thought the `data-js-hook` was for anything that was purely there for use by the JS, but `data-state` is fine with me too if that fits better. I'll let you'll sort it out in my absence :-)

File: cfgov/unprocessed/js/molecules/Expandable.js (hunk 4 above)
Comment (anselmbradford) [inline cfgov/unprocessed/js/molecules/Expandable.js]: What would this look like?

**Comment by sebworks** [comment]
:+1: . We can address the `Object. defineProperty` comment when you get back. @KimberlyMunoz can you review this again? If it's ok then I'll go ahead and merge it.

**Comment by KimberlyMunoz** [comment]
New stuff needs to be pulled in.

We should probably take note of the weird focusing issues. 
image

**Comment by sebworks** [comment]
@KimberlyMunoz, this should be good to go now. I've created and [issue](https://github.com/cfpb/cfgov-refresh/issues/1161) for the focus bug.

**Comment by KimberlyMunoz** [comment]
Thanks for the reminder. I'll pull it down right now

File: CHANGELOG.md [hunk 5]
```diff
@@ -94,7 +94,14 @@ Given the `MAJOR.MINOR.PATCH` pattern, here is how we decide to increment:
 - Added WP Import Data Contact processor
 - Added templates and CSS for the Adding Sidebar Breakout organism.
 - Added cf-tables and tables molecule
+<<<<<<< HEAD
```
Comment (KimberlyMunoz) [inline CHANGELOG.md]: Can we remove the conflict stuff?

File: CHANGELOG.md [hunk 6]
```diff
@@ -157,6 +164,8 @@ Given the `MAJOR.MINOR.PATCH` pattern, here is how we decide to increment:
 - Upgrade wagtail to 1.2
 - Cleaned up and rebuilt the secondary nav to reduce complexity and fix bugs
 - Routed landing page type related molecules and organisms to use jinja2/v1/_includes/ template locations
+- Reduced padding on expandables per direction of design.
+- Hid cues on expandables when JS is turned off.
```
Comment (KimberlyMunoz) [inline CHANGELOG.md]: "Hide cues"

File: CHANGELOG.md [hunk 7]
```diff
@@ -321,12 +330,6 @@ Given the `MAJOR.MINOR.PATCH` pattern, here is how we decide to increment:
 ## 3.0.0-2.1.0 - 2015-08-05
 
 ### Added
-- Added `sub-pages/civil-penalty-fund-allocation-schedule/` page.
```
Comment (KimberlyMunoz) [inline CHANGELOG.md]: Why were these deleted?

Comment (sebworks) [inline CHANGELOG.md]: I believe these were in the log twice.

Comment (KimberlyMunoz) [inline CHANGELOG.md]: :ok_hand:

File: CHANGELOG.md (hunk 6 above)
Comment (anselmbradford) [inline CHANGELOG.md]: Either is fine, but changelog is generally past tense.

Comment (KimberlyMunoz) [inline CHANGELOG.md]: :ok_hand:

File: cfgov/unprocessed/css/molecules/expandable.less [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+    &_content {
+        height: 0;
+        border-bottom: 1px solid @gray-50;
+        overflow: hidden;
+        position: relative;
+        background: @white;
+
+        &-animated {
+            position: relative;
```
Comment (jimmynotjim) [inline cfgov/unprocessed/css/molecules/expandable.less]: This is overwriting L:132. Also, all styles on `_content-animated` should be in one place.

File: cfgov/unprocessed/css/molecules/expandable.less [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+// .expandable_label
+@expandable_label-text:         #000000;
+
+// .expandable_link
+@expandable_link-text:          #046b99; // $color-primary-alt-darkest
+
+// Sizing variables
+
+// .expandable_link
+@expandable_label-font-size:    16px;
```
Comment (jimmynotjim) [inline cfgov/unprocessed/css/molecules/expandable.less]: Why use this if it's the default font size?

File: cfgov/unprocessed/css/molecules/expandable.less [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    &_header-right {
+        float: right;
+        width: 22%;
+        text-align: right;
+    }
+
+    &_label {
+        font-size: unit(@expandable_label-font-size / @base-font-size-px, em);
```
Comment (jimmynotjim) [inline cfgov/unprocessed/css/molecules/expandable.less]: `color` should come first to match Front-end guide

File: cfgov/unprocessed/js/routes/browse-basic/index.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+'use strict';
+
+// Required polyfills for <IE8.
+require( '../../modules/polyfill/query-selector' );
+
+// List of organisms used.
+var Expandable = require( '../../molecules/Expandable' );
+var ExpandableGroup = require( '../../organisms/ExpandableGroup' );
+
+var expandables = document.querySelectorAll( '.expandable-prototypes .m-expandable' );
```
Comment (jimmynotjim) [inline cfgov/unprocessed/js/routes/browse-basic/index.js]: Why nest in `.expandable-prototypes` and not just target `.m-expandable` directly? Same question for groups on L:21

File: cfgov/unprocessed/css/molecules/expandable.less (hunk 9 above)
Comment (anselmbradford) [inline cfgov/unprocessed/css/molecules/expandable.less]: [cf-expandables](https://github.com/cfpb/cf-expandables/blob/gh-pages/src/cf-expandables.less#L48-L67) has a color customization setting for the label and text, and a size customization setting for the text size. I added the label text size for consistency. Would:

```
@expandable_label-font-size:     @base-font-size-px;
```

…be better?

File: cfgov/unprocessed/js/routes/browse-basic/index.js (hunk 11 above)
Comment (anselmbradford) [inline cfgov/unprocessed/js/routes/browse-basic/index.js]: Because expandables can appear on their own or within expandable groups. When they appear in an expandable group, the expandable group handles initialization of the expandables it contains. So this scopes the query only to the lone expandables. In L:21 `.expandable-group-prototypes` could be removed, but is provided just for consistency.

Comment (jimmynotjim) [inline cfgov/unprocessed/js/routes/browse-basic/index.js]: Ah, thanks for the explanation, I should have realized that and was just going to fast. :ok_hand:

File: cfgov/unprocessed/css/molecules/expandable.less (hunk 9 above)
Comment (jimmynotjim) [inline cfgov/unprocessed/css/molecules/expandable.less]: Yeah, I'd point it at the base var so that if that changes, it changes here as well.

Comment (anselmbradford) [inline cfgov/unprocessed/css/molecules/expandable.less]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/9672717cdb0c17aa37e8f71d8fbe499aace61224

File: cfgov/unprocessed/css/molecules/expandable.less (hunk 8 above)
Comment (anselmbradford) [inline cfgov/unprocessed/css/molecules/expandable.less]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/d0400b6bf136901127a997c203bc63bd1de105ce

File: cfgov/unprocessed/css/molecules/expandable.less (hunk 10 above)
Comment (anselmbradford) [inline cfgov/unprocessed/css/molecules/expandable.less]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/db8aa458b54a98cbfce49223aebeea7e336afe1e

File: cfgov/jinja2/v1/_includes/molecules/expandable.html (hunk 3 above)
Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: Did this get sorted?

File: cfgov/unprocessed/js/molecules/Expandable.js (hunk 4 above)
Comment (jimmynotjim) [inline cfgov/unprocessed/js/molecules/Expandable.js]: Did this one get sorted as well?

Comment (anselmbradford) [inline cfgov/unprocessed/js/molecules/Expandable.js]: See https://github.com/cfpb/cfgov-refresh/pull/1153#issuecomment-157188781

File: cfgov/jinja2/v1/_includes/molecules/expandable.html (hunk 3 above)
Comment (anselmbradford) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: What do you'll want to do for this one?

Comment (sebworks) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: I think `data-state=expanded` make more sense but its up to you

File: cfgov/unprocessed/js/molecules/Expandable.js (hunk 4 above)
Comment (sebworks) [inline cfgov/unprocessed/js/molecules/Expandable.js]: So I was thinking the we use `Object. defineProperty` the same way we are using it [here](https://github.com/cfpb/cfgov-refresh/blob/flapjack/cfgov/unprocessed/js/modules/util/aria-state.js). This way you don't have the setters.

File: cfgov/jinja2/v1/_includes/molecules/expandable.html (hunk 3 above)
Comment (anselmbradford) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/2d3f92646844c768a3580edb6f54ddc548e7ac77

**Comment by sebworks** [comment]
:+1:
